### PR TITLE
Install CLR palette on MacOS

### LINKED ISSUE #28: Install CLR palette on MacOS
Add a feature to install a CLR color file on MacOS.

All that needs to be done is to make a copy of the CLR file into `~/Library/Colors/` folder. Please do not call `bash` or `zsh` commands, use a Python call to copy a file to a specific folder.

Sample folder output:

```
❯ ls -ltra ~/Library/Colors/
total 320
-rw-r--r--    1 adrian  staff    444 24 Mar  2022 Scrivener.clr
-rw-r--r--@   1 adrian  staff   2299 17 Aug 18:43 AWS 2021 Q1 Extract.clr
-rw-r--r--@   1 adrian  staff  61131 18 Aug 15:38 Copic 358 (by Adrian).clr
-rw-r--r--    1 adrian  staff    146 18 Aug 16:44 Copic by Yu Lei.clr
-rw-r--r--@   1 adrian  staff  61131 18 Aug 16:44 Copic 358 (by Yu Lei).clr
-rw-r--r--@   1 adrian  staff   5687 20 Aug 10:52 Solarised.clr
-rw-r--r--@   1 adrian  staff   7094 23 Aug 10:09 Doodle.clr
-rw-r--r--    1 adrian  staff   1427 28 Sep 19:55 Whimsical.clr
-rw-r--r--    1 adrian  staff   6904  4 Nov 10:25 NSColorPanelSwatches.plist
drwx------+  11 adrian  staff    352  4 Nov 10:25 .
drwx------@ 114 adrian  staff   3648 30 Nov 18:26 ..
```

### PR DESCRIPTION
**Pull request by nguyenhieu123**
Closes #28

This Pull Request:

1. Copy .clr file to Color folder in macos.
2. Add function "_copy_file_macos" in harmony/to_clr_convertion/commands.py".
3. Change function "FromTxtCommandUtils.convert_from_txt_file()" from return output none to return file_path (string).

The problem of issue is that when running the command to create .clr file, the file is created in the project's folder, the user has to move the file to the Colors folder of macos. The idea is that when running command create the .clr file, I make a copy and move it to the Colors folder of macos. I change the response of function  "FromTxtCommandUtils.convert_from_txt_file()" from return none to  return string which returns the file path of the newly created .clr file. Function "_copy_file_macos" in "harmony/to_clr_convertion/commands.py" has the input parameter (new file path of .clr file), it copy the .clr file to Colors  folder of macos.

### CODE REVIEW

**Comment by AdrianSimionov** [comment]
Can you please describe in words the functionality you are adding? Also, make sure the added functionality is documented in  the README file.

**Comment by nguyenhieu123** [comment]
Yes, I will add it

**Comment by AdrianSimionov** [comment]
What I meant by describing in words, is to describe the functionality in plain English, as below:

Does this change break the previous functionality where a CLR file was created locally? Does it change it? How is changed?
Does this change add an argument so that when the argument is used, the file is also copied to the MacOS Library folder?
If such an argument is present, is the behavior described properly in the documentation?

File: harmony/to_clr_convertion/commands.py [hunk 1]
```diff
@@ -12,6 +16,17 @@ def txt2clr(
     suffix: str = core.CommonArguments.suffix,
 ):
     """Command to convert a text file into a ".clr" file"""
-    FromTxtCommandUtils(CLRWriting(), suffix, generate_names).convert_from_txt_file(
-        file_path
-    )
+    file_path_clr = FromTxtCommandUtils(
+        CLRWriting(), suffix, generate_names
+    ).convert_from_txt_file(file_path)
+    _copy_file_macos(file_path_clr)
```
Comment (xlurio) [inline harmony/to_clr_convertion/commands.py:22]: This line is breaking the functionality of the command. This command is only for converting "Harmony" files into `.clr` files. "Installing" `.clr` should be placed in a new command. The new command should also raise an Exception in case the file is not a `.clr` file (doesn't have the `04 0b 73 74 72 65 61 6d 74 79 70 65 64` signature byte string at the beginning of the file) or the current OS is not MacOS, since this functionality is exclusive to Apple operating systems

**Review by xlurio — CHANGES_REQUESTED** [review]
Please, also review the name of your branch. We shall use the git flow naming conventions (https://dev.to/couchcamote/git-branching-name-convention-cch).

File: harmony/to_clr_convertion/commands.py (hunk 1 above)
Comment (xlurio) [reply harmony/to_clr_convertion/commands.py:22]: In respect to https://github.com/verocloud/color-sorting-cli/issues/28#issuecomment-1480297615, instead of creating a new command we will add a new flag option (which value will be `False` by default). The option must have a OS checking which will raise an error in case it is used in a OS different from MacOS.

**Comment by nguyenhieu123** [comment]
Yes, I understand, I will fixed it.

File: harmony/to_clr_convertion/commands.py (hunk 1 above)
Comment (nguyenhieu123) [reply harmony/to_clr_convertion/commands.py:22]: Hello xlurio, Do we need created the new flag? I think we can checking os in command logic, if it is macos then i will copy file to Colors library and if not, We run command normally.

**Comment by AdrianSimionov** [comment]
Please add changes here.

**Comment by nguyenhieu123** [comment]
> Please add changes here.

I created new branch. Can I created merge branch to this branch.

**Comment by nguyenhieu123** [comment]
This Pull Request:

1. Copy .clr file to Color folder in macos.
2. Add function "copy_file_macos" , "checking_os", "check_os_and_copy_file" in "hamony/to_clr_convertion/utils"
3. Add new flag "--mac_os" default is False in "txt2clr" command
4. Add test for write .clr file for macos

The problem of issue is that when running the command to create .clr file, the file is created in the project's folder, the user has to move the file to the Colors folder of macos. The idea is that when running command create the .clr file, I make a copy and move it to the Colors folder of macos. I add the new flag "--macos" in command write .clr file to run function copy file to Colors folder if it's os is macos .

File: tests/writing_strategies_tests/test_clr_writing.py [hunk 2]
```diff
@@ -82,3 +84,53 @@ def _when_colors_are_passed_writing_as_clr(
         os.remove(temporary_file)
 
         return colors_file_content
+
+    def test_write_as_clr_file_with_macos(self, monkeypatch) -> None:
```
Comment (xlurio) [inline tests/writing_strategies_tests/test_clr_writing.py:88]: In Harmony API, every entity should have its own test module, and you are using the module that tests the `CLRWriting` class to test other entity. Please, follow the pattern and write your tests in a separated module.
For the file you gonna use for the test, you can convert a Harmony file into a `.clr` file and place it into `tests/resources` folder. Then you copy the file into a [temporary directory](https://docs.python.org/3.8/library/tempfile.html#tempfile.TemporaryDirectory) and work the test from there.
Also, please [setup a Docker container](https://dockerlabs.collabnix.com/beginners/dockerfile/lab_dockerfile_python.html) for the tests, I don't think nobody wants to have a test messing around with them OSs everytime the tests are runned. Stay free for calling for help, I know Docker documentation is not one of the best.

**Review by xlurio — CHANGES_REQUESTED** [review]

File: harmony/commands.py [hunk 3]
```diff
@@ -13,17 +14,18 @@ class FromTxtCommandUtils:
     def __init__(
         self,
         writing_strategy: interfaces.WritingStrategy,
-        suffix: str,
```
[the remark this answers is not in the document]
Comment (nguyenhieu123) [reply harmony/commands.py:16]: Since the lint test raise "too many argument in function", I add the the new argument in kwags.

File: harmony/to_clr_convertion/utils.py [hunk 4]
```diff
@@ -29,3 +36,35 @@ def is_clr_color_component_near_zero(component_as_decimal: float) -> bool:
     return core.are_almost_equal(
         component_as_decimal, 0, core.FloatComparisonTolerance.THREE_DECIMAL_PLACES
     )
+
+
+def copy_file_macos(file_path_clr) -> None:
```
[the remark this answers is not in the document]
Comment (nguyenhieu123) [reply harmony/to_clr_convertion/utils.py:41]: Since the lint test raise "too many argument in function", I add the the new argument in kwags. When i get argument keywords, i set it default value

[the remark this answers is not in the document]
Comment (nguyenhieu123) [reply harmony/to_clr_convertion/utils.py:41]: I will put two function in class but when i put it into term service, lint's checking raise line is too long, the number of line in service is limited

File: tests/writing_strategies_tests/test_clr_writing.py (hunk 2 above)
Comment (nguyenhieu123) [reply tests/writing_strategies_tests/test_clr_writing.py:88]: Can you guide me or give document to package test documentation into docker file.

File: harmony/commands.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (xlurio) [reply harmony/commands.py:16]: You can disable this error for `__init__` methods. Do like this:
```
def __init__( # pylint: disable=too-many-arguments
        ...
):
    ...
```

Some classes may have a lot of attributes and, because of that, `__init__` may have a lot of arguments. Other functions and methods should have no more than three arguments.

File: harmony/commands.py [hunk 5]
```diff
@@ -13,17 +14,18 @@ class FromTxtCommandUtils:
...
         must_generate_color_names: bool = False,
+        **kwargs,
     ) -> None:
         self.__writing_strategy = writing_strategy
         self.__must_generate_color_names = must_generate_color_names
-        self.__suffix = suffix
+        self.__suffix = str(kwargs.get("suffix", ""))
+        self.__mac_os = bool(kwargs.get("mac_os", False))
```
[the remark this answers is not in the document]
Comment (xlurio) [reply harmony/commands.py:23]: As I already mentioned in the other comment, the `too-many-arguments` can be ignored for `__init__` method.
As for the default value you could do just `kwargs["mac_os"]` since a case where the `__init__` does not receive this argument is a programming error.

File: harmony/to_clr_convertion/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (xlurio) [reply harmony/to_clr_convertion/utils.py:41]: You can create a new module called `io_services` or something, just don't use the term `utils` unless it attends the said terms.

File: tests/writing_strategies_tests/test_clr_writing.py (hunk 2 above)
Comment (xlurio) [reply tests/writing_strategies_tests/test_clr_writing.py:88]: Well there is the [official documentation](https://docs.docker.com/language/python/build-images/), but I think it's easier for us to build it together. Be in Discord tomorrow at 14:00 UTC and we will work on it.

File: harmony/commands.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nguyenhieu123) [reply harmony/commands.py:16]: thank you, I will follow it

File: tests/writing_strategies_tests/test_clr_writing.py (hunk 2 above)
Comment (nguyenhieu123) [reply tests/writing_strategies_tests/test_clr_writing.py:88]: Great, See you at 14:00 UTC

**Comment by nguyenhieu123** [comment]
The pull request:

1. Change the tag from --macos to --install.
2. Claim the copy_file and check_os function to class and change it to "io_service.py".
3. Write the docker file for test, and change the test for copy file function.
4. Remove "kwangs" in argument of some function

File: harmony/core/__init__.py [hunk 6]
```diff
@@ -12,6 +12,7 @@
     ImageModesForPIL,
     Resources,
     SortingStrategyName,
+    Txt2ClrArguments,
```
Comment (xlurio) [inline harmony/core/__init__.py:15]: the `core` module is only for the **shared API**, `Txt2ClrArguments` is only used into `to_clr_convertion` thus it should be in `to_clr_convernio.constants`

File: harmony/core/constants.py [hunk 7]
```diff
@@ -131,3 +131,14 @@ class SortingStrategyName(str, Enum):
     STEP: str = "step"
     ALTERNATED_STEP: str = "stepalt"
     HILLBERT: str = "hillbert"
+
+
+class Txt2ClrArguments:
```
Comment (xlurio) [inline harmony/core/constants.py:136]: https://github.com/verocloud/color-sorting-cli/pull/77/files#r1157855573

**Review by xlurio — CHANGES_REQUESTED** [review]
Just a few changes left!

File: harmony/core/constants.py [hunk 8]
```diff
@@ -131,3 +131,14 @@ class SortingStrategyName(str, Enum):
...
+    """Argument for txt2clr command"""
+
+    install: bool = typer.Option(
+        False,
+        "--install",
+        "-i",
+        help="Save the file .clr to Colors folder in macos",
```
Comment (xlurio) [inline harmony/core/constants.py:143]: Please change the description to `"Store the '.clr' as a color palette in MacOS"` and take care with upper and lower cases as this text will be showed to the user.

File: harmony/to_clr_convertion/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (xlurio) [reply harmony/to_clr_convertion/utils.py]: Item **2** is still pending

File: harmony/to_clr_convertion/io_services.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        self.__flag = flag
+
+    def copy_file_macos(self) -> None:
+        """Copy file .clr to Colors folder in macos"""
+
+        source = self.__file_path
+        destination = os.path.expanduser("~") + "/Library/Colors/"
+        try:
+            shutil.copy(source, destination)
+            rich.print("[green]File is copied to Colors folder successfully!")
```
Comment (xlurio) [inline harmony/to_clr_convertion/io_services.py:24]: Please, avoid outputting messages outside commands. Add a new attribute to the class which will become `True` in case the `.clr` file was installed as a color palette and move the `rich.print()` line to the command.

File: harmony/commands.py [hunk 10]
```diff
@@ -44,9 +46,10 @@ def convert_from_txt_file(
                 file_path, self.__writing_strategy.EXTENSION
             )
         )
-        return core_services.PathGenerator(self.__suffix).get_path_with_extension(
-            file_path, self.__writing_strategy.EXTENSION
-        )
+        file_path_str = core_services.PathGenerator(
+            self.__suffix
+        ).get_path_with_extension(file_path, self.__writing_strategy.EXTENSION)
+        utils.check_os_and_copy_file(self.__mac_os, file_path_str)
```
[the remark this answers is not in the document]
Comment (xlurio) [reply harmony/commands.py]: Item **2** still pending
